### PR TITLE
Update developer guide to recommend no default stream parameter.

### PR DESCRIPTION
**Pull request by bdice**
Updates the developer guide to address #9854. With this change, any new or refactored detail APIs should have no default stream parameter. We may resolve that issue or perhaps leave it open until a majority of code has no default stream.

### CODE REVIEW

File: cpp/docs/DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -362,14 +363,14 @@ void external_function(...);
...
 // cpp/include/cudf/detail/header.hpp
 namespace detail{
-void external_function(..., rmm::cuda_stream_view stream = rmm::cuda_stream_default)
+void external_function(..., rmm::cuda_stream_view stream)
 } // namespace detail
 
 // cudf/src/implementation.cpp
 namespace detail{
-    // defaulted stream parameter
+    // use stream parameter in detail implementation
```
Comment (ttnghia) [inline cpp/docs/DEVELOPER_GUIDE.md:371]:
```suggestion
    // Use stream parameter in detail implementation.
```

File: cpp/docs/DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -362,14 +363,14 @@ void external_function(...);
...
     void external_function(..., rmm::cuda_stream_view stream){
-        // implementation uses stream w/ async APIs
+        // implementation uses stream with async APIs
```
Comment (ttnghia) [inline cpp/docs/DEVELOPER_GUIDE.md:373]: Sorry that I'm a bit OCD about comments: First letter capitalized, and punctuation at the end.
```suggestion
        // Implementation uses stream with async APIs.
```

File: cpp/docs/DEVELOPER_GUIDE.md [hunk 3]
```diff
@@ -379,7 +380,7 @@ namespace detail{
 
 void external_function(...){
     CUDF_FUNC_RANGE(); // Auto generates NVTX range for lifetime of this function
```
Comment (bdice) [inline cpp/docs/DEVELOPER_GUIDE.md:382]:
```suggestion
    CUDF_FUNC_RANGE(); // Generates NVTX range for lifetime of this function.
```

**Review by ttnghia — APPROVED** [review]

File: cpp/docs/DEVELOPER_GUIDE.md [hunk 4]
```diff
@@ -347,9 +347,10 @@ implemented using asynchronous APIs on the default stream (e.g., stream 0).
 
 The recommended pattern for doing this is to make the definition of the external API invoke an
 internal API in the `detail` namespace. The internal `detail` API has the same parameters as the
-public API, plus a `rmm::cuda_stream_view` parameter at the end defaulted to
-`rmm::cuda_stream_default`. The implementation should be wholly contained in the `detail` API
-definition and use only asynchronous versions of CUDA APIs with the stream parameter.
+public API, plus a `rmm::cuda_stream_view` parameter at the end with no default value. The public
```
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/docs/DEVELOPER_GUIDE.md]: This is a great suggestion! Applied in b3052cd.

**Review by mythrocks — APPROVED** [review]

File: cpp/docs/DEVELOPER_GUIDE.md [hunk 5]
```diff
@@ -362,14 +364,14 @@ void external_function(...);
...
 // cpp/include/cudf/detail/header.hpp
 namespace detail{
-void external_function(..., rmm::cuda_stream_view stream = rmm::cuda_stream_default)
+void external_function(..., rmm::cuda_stream_view stream)
 } // namespace detail
 
 // cudf/src/implementation.cpp
 namespace detail{
-    // defaulted stream parameter
+    // Use stream parameter in detail implementation.
```
Comment (davidwendt) [inline cpp/docs/DEVELOPER_GUIDE.md:372]:
```suggestion
    // Use the stream parameter in the detail implementation.
```

File: cpp/docs/DEVELOPER_GUIDE.md [hunk 6]
```diff
@@ -362,14 +364,14 @@ void external_function(...);
...
     void external_function(..., rmm::cuda_stream_view stream){
-        // implementation uses stream w/ async APIs
+        // Implementation uses stream with async APIs.
```
Comment (davidwendt) [inline cpp/docs/DEVELOPER_GUIDE.md:374]:
```suggestion
        // Implementation uses the stream with async APIs.
```

File: cpp/docs/DEVELOPER_GUIDE.md [hunk 7]
```diff
@@ -378,8 +380,8 @@ namespace detail{
 } // namespace detail
 
 void external_function(...){
-    CUDF_FUNC_RANGE(); // Auto generates NVTX range for lifetime of this function
-    detail::external_function(...);
+    CUDF_FUNC_RANGE(); // Generates NVTX range for lifetime of this function.
```
Comment (davidwendt) [inline cpp/docs/DEVELOPER_GUIDE.md:383]:
```suggestion
    CUDF_FUNC_RANGE(); // Generates an NVTX range for the lifetime of this function.
```

**Comment by bdice** [comment]
@gpucibot merge

**Pull request merged**